libs/core/src/lib/content-density/classes/content-density-consumer.class.ts

```diff
@@ -0,0 +1,140 @@
+    private _condensedSubscription: Subscription;
+    /** @hidden */
+    private _cozySubscription: Subscription;
+
+    /** @hidden */
+    private _callCallbacks$ = new BehaviorSubject<void>(undefined);
+
+    /** @hidden */
+    constructor(
+        initialValue: ContentDensityMode,
+        dataSource: Observable<ContentDensityMode>,
+        private destroy$: Observable<void>,
+        private changeDetectorRef: ChangeDetectorRef
+    ) {
+        super(initialValue);
+        const source$ = dataSource.pipe(distinctUntilChanged(), takeUntil(this.destroy$));
+        combineLatest([source$, this._callCallbacks$])
+            .pipe(
+                tap(([density]) => this._callCallbacks(density)),
+                map(([density]) => density),
+                finalize(() => this.complete())
+            )
+            .subscribe((density) => {
+                this.next(density);
+            });
+        this.isCompact$ = source$.pipe(map(isCompact));
+        this.isCozy$ = source$.pipe(map(isCozy));
+        this.isCondensed$ = source$.pipe(map(isCondensed));
+    }
+
+    /**
+     * Sync isCompact. Calls ChangeDetectorRef.markForCheck() on update
+     */
+    get isCompact(): boolean {
+        if (!this._compactSubscription) {
+            this._compactSubscription = this._listenToChanges(this.isCompact$, (v) => (this._isCompact = v));
+        }
+        return this._isCompact;
+    }
+
+    /**
+     * Sync isCondensed. Calls ChangeDetectorRef.markForCheck() on update
+     */
+    get isCondensed(): boolean {
+        if (!this._condensedSubscription) {
+            this._condensedSubscription = this._listenToChanges(this.isCondensed$, (v) => (this._isCondensed = v));
+        }
+        return this._isCondensed;
+    }
+
+    /**
+     * Sync isCondensed. Calls ChangeDetectorRef.markForCheck() on update
+     */
+    get isCozy(): boolean {
+        if (!this._cozySubscription) {
+            this._cozySubscription = this._listenToChanges(this.isCozy$, (v) => (this._isCozy = v));
+        }
+        return this._isCozy;
+    }
+
+    /**
+     * Pass consumers callback function, or configuration object to be called on density change
+     */
+    consume(
+        ...consumerConfigs: Array<ContentDensityConsumerTarget | ContentDensityCallbackFn>
+    ): ContentDensityObserver {
+        consumerConfigs.forEach((consumerConfig) => {
+            const callback = contentDensityCallbackFactory(consumerConfig);
+            this.callbacks.set(consumerConfig, callback);
+        });
+        this._callCallbacks$.next();
+        return this;
+    }
+
+    /**
+     * Remove consumer callback from list of callbacks
+     */
+    removeConsumer(
+        ...consumerConfigs: Array<ContentDensityConsumerTarget | ContentDensityCallbackFn>
+    ): ContentDensityObserver {
+        consumerConfigs.forEach((consumerConfig) => {
+            this.callbacks.delete(consumerConfig);
+        });
+        this._callCallbacks$.next();
```

Comment:
every time we add or remove a consumer we run all the consumers because of this emit. but in the typedocs said `to be called on density change`. is it intended?

Reply:
Hmm, now when you mentioned it, I remember that I should not call everyone, one is enough :D